feat: astro features

ematipico opened the pull request:
## Changes

Introducing **Astro features**. Astro features are features that astro has out the box, and sometimes users can opt-in. 
Few examples:
- `staticOutput` if a user uses a specific adapter, the adapter needs to be able to serve those static assets. An example is Cloudflare, which **doesn't support it**.
- `assets`: not supported by those adapters that are incompatible with node
 
___

This PR adds a new API that adapters can use when registering one. It's called `supportedFeatures`, and it's a new way to tell Astro if and when an adapter is able to support certain features that are enabled via configuration. 

The adapter can also tell Astro the kind of support:
- stable
- experimental
- unsupported
- deprecated


For now, this new API/object is not mandatory, but it will be in Astro 4.0.

The PR also updates the adapters that we maintain, and it adds the feature map to them.

### About `assets`

There's a small caveat for the assets. Astro assets is a feature that is already turned on by default, meaning that Astro sets already a default service. This means that adapters that **don't support** assets will always get an error, which is inconvenient. I came up with a solution where we set the default **after** the adapter validation, although I don't like the solution very much, and I wonder if there's a better way to handle a case like this.

The assets feature is just an example; we might have the same problem with future features like this. What do you suggest?

## Testing

I added new unit test cases and updated the existing ones.

<!-- How was this change tested? -->
<!-- DON'T DELETE THIS SECTION! If no tests added, explain why. -->

## Docs

I will create the changelog in another PR

<!-- Could this affect a user’s behavior? We probably need to update docs! -->
<!-- If docs will be needed or you’re not sure, uncomment the next line: -->
<!-- /cc @withastro/maintainers-docs for feedback! -->

<!-- DON'T DELETE THIS SECTION! If no docs added, explain why.-->
<!-- https://github.com/withastro/docs -->

bluwy reviewed (commented):
The general validation logic looks great! Got some questions/comments about some things

bluwy commented on packages/astro/src/integrations/index.ts:
```diff
@@ -178,13 +186,183 @@ export async function runHookConfigDone({
 			});
 		}
 	}
+	// Astro doesn't set a default on purpose when loading the configuration file.
+	// Not setting a default allows to check if the adapter doesn't support the feature in case the user overrides
+	// Astro's defaults.
+	//
+	// At the end of the validation, we push the correct defaults.
+	if (typeof settings.config.image === 'undefined') {
```
I'm actually not sure if we can get the adapter config from here. It might be too late 🤔

ematipico replied:
Oh, shoot, that's old code that I have to remove :D good catch!

Princesseuh commented on packages/astro/src/@types/astro.ts:
```diff
@@ -1660,12 +1660,47 @@ export type PaginateFunction = (data: any[], args?: PaginateOptions) => GetStati
+	 * The adapter is able to serve SSR pages
+	 */
+	serverOutput?: SupportsKind;
+	/**
+	 * The adapter can emit static assets
+	 */
+	assets?: {
+		supportKind?: SupportsKind;
+		/**
+		 * Whether if this adapter deploys files in an enviroment that is Node.js compatible.
+		 *
+		 * The default services used by Astro needs Node.js.
+		 */
+		isNodeCompatible?: boolean;
+	};
```
Hmm, definitely not for `Squoosh`, especially since we plan on removing it. But for Sharp I guess it could be a thing? It does have native deps, albeit nowadays it kinda supports everything and it'll hopefully support WASM soon so it's unlikely for a platform to have a specific issue with it anymore (even Stackblitz and other web envs have it working now)

ematipico commented:
I updated the code:
- the validation doesn't throw an error, but it **logs** an error. This was making testing impossible.
- I removed the hacks I needed for deno tests
- @bluwy, I updated the code based on your suggestion; please check again
- @Princesseuh, could you please check the **adapters** and make sure the values are correct?

Princesseuh commented on packages/integrations/vercel/src/edge/adapter.ts:
```diff
@@ -27,6 +27,15 @@ function getAdapter(): AstroAdapter {
+			hybridOutput: 'Stable',
+			staticOutput: 'Stable',
+			serverOutput: 'Stable',
+			assets: {
+				supportKind: 'Unsupported',
+				isSharpCompatible: true,
+			},
```
On Vercel Edge, the only image service supported is Vercel's (when using `vercel({imageService: true})`)

Princesseuh replied:
```suggestion
			assets: {
				supportKind: 'Stable',
				isSharpCompatible: true,
				isSquooshCompatible: false,
			},
```

Princesseuh commented on .changeset/dirty-lies-cover.md:
```diff
@@ -0,0 +1,31 @@
+            hybridOutput: 'Experimental',
+            staticOutput: 'Stable',
+            serverOutput: 'Stable',
+            assets: {
+                supportKind: 'Unsupported',
+                isNodeCompatible: false,
+            },
```
Outdated?

Princesseuh commented on packages/integrations/cloudflare/src/index.ts:
```diff
@@ -24,11 +24,31 @@ export function getAdapter(isModeDirectory: boolean): AstroAdapter {
+					staticOutput: 'Unsupported',
+					serverOutput: 'Stable',
+					assets: {
+						supportKind: 'Unsupported',
+						isSharpCompatible: false,
+						isSquooshCompatible: false,
```
```suggestion
						supportKind: 'Stable',
						isSharpCompatible: false,
						isSquooshCompatible: false,
```

Every adapter supports `assets` the feature, what they don't support is the image services. Someone could write an image service powered by a supported library (or Cloudflare's Image API in this case)

Princesseuh commented on packages/integrations/netlify/src/integration-edge-functions.ts:
```diff
@@ -11,6 +11,16 @@ export function getAdapter(): AstroAdapter {
+			staticOutput: 'Stable',
+			serverOutput: 'Stable',
+			assets: {
+				supportKind: 'Unsupported',
```
```suggestion
				supportKind: 'Stable',
```

Same as Cloudflare, every adapter supports `assets`, it's the services they don't support

matthewp commented:
Should the values be lower-case? When we have string flags like this elsewhere in the code I'm pretty sure they are lower-case.

bluwy commented on packages/astro/src/integrations/index.ts:
```diff
@@ -178,6 +179,31 @@ export async function runHookConfigDone({
+							warn(
+								logging,
+								'astro',
+								`The adapter ${adapter.name} doesn't provide a feature map. From Astro 3.0, an adapter 
+		can provide a feature map. Not providing a feature map will cause an error in Astro 4.0.`
```
nit: I think the string should be a single line so it can be line-wrapped in the terminal naturally.
